extensions/typescript-language-features/src/languageFeatures/inlayHints.ts

```diff
@@ -0,0 +1,143 @@
+		/**
+		 * Start position of the span.
+		 */
+		start: number;
+		/**
+		 * Length of the span.
+		 */
+		length: number;
+	}
+
+	export interface InlineHintsRequest extends Proto.Request {
+		command: CommandTypes.ProvideInlineHints;
+		arguments: InlayHintsArgs;
+	}
+
+	export enum InlayHintKind {
+		Type = 'Type',
+		Parameter = 'Parameter',
+		Enum = 'Enum'
+	}
+
+	interface InlayHintItem {
+		text: string;
+		position: Proto.Location;
+		kind?: InlayHintKind;
+		whitespaceBefore?: boolean;
+		whitespaceAfter?: boolean;
+	}
+
+	export interface InlayHintsResponse extends Proto.Response {
+		body?: InlayHintItem[];
+	}
+
+	export interface IExtendedTypeScriptServiceClient {
+		execute<K extends keyof ExtendedTsServerRequests>(
+			command: K,
+			args: ExtendedTsServerRequests[K][0],
+			token: vscode.CancellationToken,
+			config?: ExecConfig
+		): Promise<ServerResponse.Response<ExtendedTsServerRequests[K][1]>>;
+	}
+
+	export interface ExtendedTsServerRequests {
+		'provideInlayHints': [InlayHintsArgs, InlayHintsResponse];
+	}
+
+	export namespace InlayHintKind {
+		export function fromProtocolInlayHintKind(kind: InlayHintKind): vscode.InlayHintKind {
+			switch (kind) {
+				case InlayHintKind.Parameter:
+					return vscode.InlayHintKind.Parameter;
+				case InlayHintKind.Type:
+					return vscode.InlayHintKind.Type;
+				case InlayHintKind.Enum:
+					return vscode.InlayHintKind.Other;
+				default:
+					return vscode.InlayHintKind.Other;
+			}
+		}
+	}
+}
+
+class TypeScriptInlayHintsProvider implements vscode.InlayHintsProvider {
+	public static readonly minVersion = API.v440;
+
+	constructor(
+		private readonly client: ITypeScriptServiceClient,
+		private readonly fileConfigurationManager: FileConfigurationManager
+	) { }
+
+	async provideInlayHints(model: vscode.TextDocument, range: vscode.Range, token: vscode.CancellationToken): Promise<vscode.InlayHint[]> {
+		const filepath = this.client.toOpenedFilePath(model);
+		if (!filepath) {
+			return [];
+		}
+
+		await this.fileConfigurationManager.ensureConfigurationForDocument(model, token);
+
+		if (!this.someInlayHintsEnabled(model)) {
+			return [];
+		}
+
+		const start = model.offsetAt(range.start);
+		const length = model.offsetAt(range.end) - start;
+
+		const response = await (this.client as ExperimentalProto.IExtendedTypeScriptServiceClient).execute('provideInlayHints', { file: filepath, start, length }, token);
+		if (response.type !== 'response' || !response.success || !response.body) {
+			return [];
+		}
+
+		return response.body.map(hint => {
+			const result = new vscode.InlayHint(
+				hint.text,
+				Position.fromLocation(hint.position),
+				hint.kind && ExperimentalProto.InlayHintKind.fromProtocolInlayHintKind(hint.kind)
+			);
+			result.whitespaceBefore = hint.whitespaceBefore;
+			result.whitespaceAfter = hint.whitespaceAfter;
+			return result;
+		});
+	}
+
+	private someInlayHintsEnabled(model: vscode.TextDocument) {
```

Comment:
Let's cache this value and update it when the configuration changes. Reading the configuration every time hints are requested is expensive

Reply:
Alternatively, you can unregister the provider entirely when no hints are enabled. The JSDoc completion provider shows something similar: https://github.com/microsoft/vscode/blob/192c817fd350bcbf3caecae22a45ec39bae78516/extensions/typescript-language-features/src/languageFeatures/jsDocCompletions.ts#L130

Reply:
Done.